packages/react/src/components/cookieConsent/modal/Modal.tsx

```diff
@@ -0,0 +1,74 @@
+import React, { useContext, useEffect, useState } from 'react';
+import ReactDOM from 'react-dom';
+import { VisuallyHidden } from '@react-aria/visually-hidden';
+
+import classNames from '../../../utils/classNames';
+import styles from '../CookieConsent.module.scss';
+import { CookieConsentContext, useCookieConsentUiTexts, useFocusShift } from '../CookieConsentContext';
+import { Content } from '../content/Content';
+import { useEscKey } from '../useEscKey';
+
+export function Modal(): React.ReactElement | null {
+  const cookieConsentContext = useContext(CookieConsentContext);
+  const hasOnlyRequiredConsents =
+    !cookieConsentContext.content.optionalCookies || cookieConsentContext.content.optionalCookies.groups.length === 0;
+  const shouldShowModal = !hasOnlyRequiredConsents && !cookieConsentContext.hasUserHandledAllConsents();
+  const [isModalInitiallyShown] = useState<boolean>(shouldShowModal);
+  const [popupTimerComplete, setPopupTimerComplete] = useState<boolean>(false);
+  const popupDelayInMs = 500;
+  // if hasUserHandledAllConsents() was false at first and then later true, user must have saved consents.
+  const showScreenReaderSaveNotification = isModalInitiallyShown && !shouldShowModal;
+  const { settingsSaved } = useCookieConsentUiTexts();
+  const containerId = 'HdsCookieConsentContainer';
+  const getContainerElement = (): HTMLElement | null => document.getElementById(containerId);
+  const [isDomReady, setIsDomReady] = useState<boolean>(false);
+
+  useEffect(() => {
+    setTimeout(() => setPopupTimerComplete(true), popupDelayInMs);
+  }, []);
+
+  useEffect(() => {
+    const containerElement = getContainerElement();
+    if (shouldShowModal && !isDomReady) {
+      if (!containerElement) {
+        const htmlContainer = document.createElement('div');
+        htmlContainer.setAttribute('id', containerId);
+        htmlContainer.setAttribute('data-testid', 'html-cookie-consent-container');
+        document.body.insertBefore(htmlContainer, document.body.firstChild);
+      }
+      setIsDomReady(true);
+    } else if (!shouldShowModal && isDomReady) {
+      if (containerElement) {
+        document.body.removeChild(containerElement);
+      }
+      setIsDomReady(false);
+    }
+  }, [shouldShowModal, isDomReady, setIsDomReady]);
+
+  useEscKey(useFocusShift());
```

Comment:
There's quite much happening here.

1. `useEscKey` is a useEffect hook, which will be run after every render. Is it necessary to run it after every render? What the hook does is that it adds an event listener for the 'esc' key. When esc is pressed, then the hook will run the callback function that has been given to it in arguments. The hook has also a return value, which will remove the event listener after the component unmounts, and also before each render. So always when the component renders or re-renders, the hook will first remove the previous event listener, and add a new event listener. That is quite much for every render of this modal. Would it be possible/reasonable to add the event listener only once when the component mounts, and then remove it when it unmounts?

2. The callback for the `useEscKey` is created by calling another hook function, called `useFocusShift`. The `useFocusShift` will return a function where focus will be forced on the cookie consent `focusTargetSelector`. So all in all, when you press an 'esc' key, the element `focusTargetSelector` will be focused. There's quite many react hooks stacked here. The `useFocusShift` uses yet another react hook, called `useCookieConsentContent`, and then returns a function, in which function `forceFocusToElement` will be called.  The `useCookieConsentContent` uses yet another react hook called `useContext(CookieConsentContext)`, to get content, which has attribute `focusTargetSelector`. Im not sure if there's too many hooks stacked, which makes following and understanding the code little harder. Though then again, some of those hooks are used in multiple places, which will reduce code repetition. The `useFocusShift` hook seem to be used only inside the Modal.tsx component, so would it be good idea to move it there? It could be there inlined or renamed with a more specific meaning.

Reply:
1. Fixed resetting the document.addListener in each render. The passed callback is now a a cached with useCallback. 

But useAccordion returns new callbacks on each render, so the callback used with useEsc also changes everytime. So there the listener is added/removed on each render.

2. UseEscKey is meant to be used also in other HDS-components (there are comments in other components about creating a hook)

Yes there a lot of hooks. One adds the esc listener, one creates a function that moves the focus, one that gets the selector for the element to move the focus to.

Selector has to be passed as a prop to the component, so it has to be in the context. Hook is the easiest way to get. It could be a function too (I guess), but the function would be identical.

FocusSetter is a hook, because it uses the context hook. Making it a non-hook is impossible I think. Now it also caches the callback, so document.listener is not reset on each render.

So all hooks would be same functions anyway, if they were stripped out of their "hook status".